private/tools/java/rules/jvm/external/jar/AddJarManifestEntry.java

```diff
@@ -0,0 +1,171 @@
+package rules.jvm.external.jar;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.jar.Attributes;
+import java.util.jar.JarFile;
+import java.util.jar.JarOutputStream;
+import java.util.jar.Manifest;
+import java.util.zip.CRC32;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+/*
+ * A class that will add an entry to the manifest and keep the same modification
+ # times of the jar entries.
+ */
+public class AddJarManifestEntry {
+
+  public static final long DEFAULT_TIMESTAMP =
+      LocalDateTime.of(2010, 1, 1, 0, 0, 0)
+          .atZone(ZoneId.systemDefault())
+          .toInstant()
+          .toEpochMilli();
+
+  public static void verboseLog(String logline) {
+    // To make this work you need to add 'use_default_shell_env = True' to the
+    // rule and specify '--action_env=RJE_VERBOSE=true' to the bazel build command.
+    if (System.getenv("RJE_VERBOSE") != null) {
+      System.out.println(logline);
+    }
+  }
+
+  public static void main(String[] args) throws IOException {
+    Path out = null;
+    Path source = null;
+    String manifestEntry = null;
+
+    for (int i = 0; i < args.length; i++) {
+      switch (args[i]) {
+        case "--output":
+          out = Paths.get(args[++i]);
+          break;
+
+        case "--manifest-entry":
+          manifestEntry = args[++i];
+          break;
+
+        case "--source":
+          source = Paths.get(args[++i]);
+          break;
+
+        default:
+          throw new IllegalArgumentException("Unable to parse command line: " + Arrays.toString(args));
+      }
+    }
+
+    Objects.requireNonNull(source, "Source jar must be set.");
+    Objects.requireNonNull(out, "Output path must be set.");
+
+    byte[] buffer = new byte[2048];
+    int bytesRead;
+    boolean manifestUpdated = false;
+
+    try (InputStream fis = Files.newInputStream(source);
+         ZipInputStream zis = new ZipInputStream(fis)) {
+
+      try (OutputStream fos = Files.newOutputStream(out);
```

Comment:
You can safely merge these two `try` blocks into a single one.

Reply:
Yeah, these can be merged but it I thought keeping them separate made the blocks more clear.